### PR TITLE
Add when-let* and when-some*

### PR DESCRIPTION
**Pull request by griffinlacek**
Found myself thinking these would be nice for bindings that all need to be present before taking an action.

### CODE REVIEW

**Comment by RutledgePaulV** [comment]
Sounds good to me. I'd probably implement the if-let* / if-some* versions as well and then re-implement when-let* and when-some* on top of those.

File: test/missing/core_test.clj [hunk 1]
```diff
@@ -522,3 +522,56 @@
     (is (= () (merge-sort [[] []])))
     (is (= '(1 2 3 4 5 6) (merge-sort colls)))
     (is (= '(6 5 4 3 2 1) (merge-sort (flip compare) (map reverse colls))))))
+
+(deftest when-some-test
+  (let [atom (atom {})]
+    (when-some* [a 1
```
Comment (RutledgePaulV) [inline test/missing/core_test.clj]: You don't really need to use an atom for these tests, can just assert the return value.

**Comment by griffinlacek** [comment]
> Sounds good to me. I'd probably implement the if-let* / if-some* versions as well and then re-implement when-let* and when-some* on top of those.

I was curious about these from a discussion I saw elsewhere. Would you implement the ifs as though all bindings must evaluate to true? Seems like if it was in a similar vein to the when macros you'd need to evaluate each binding as its own if statement which would be hard to manage ifs and elses.

**Comment by RutledgePaulV** [comment]
> > Sounds good to me. I'd probably implement the if-let* / if-some* versions as well and then re-implement when-let* and when-some* on top of those.
> 
> I was curious about these from a discussion I saw elsewhere. Would you implement the ifs as though all bindings must evaluate to true? Seems like if it was in a similar vein to the when macros you'd need to evaluate each binding as its own if statement which would be hard to manage ifs and elses.

I think this would work:

```clojure
(defmacro if-let*
  "Like clojure.core/if-let except supports multiple bindings."
  ([bindings then]
   `(if-let* ~bindings ~then nil))
  ([bindings then else]
   (if (seq bindings)
     `(if-let [~@(take 2 bindings)]
        (if-let* [~@(drop 2 bindings)] ~then ~else)
        ~else)
     then)))

(defmacro if-some*
  "Like clojure.core/if-some except supports multiple bindings."
  ([bindings then]
   `(if-some* ~bindings ~then nil))
  ([bindings then else]
   (if (seq bindings)
     `(if-some [~@(take 2 bindings)]
        (if-some* [~@(drop 2 bindings)] ~then ~else)
        ~else)
     then)))

(defmacro when-let*
  "Like clojure.core/when-let except supports multiple bindings."
  [bindings & body]
  `(if-let* ~bindings (do ~@body)))

(defmacro when-some*
  "Like clojure.core/when-some except supports multiple bindings."
  [bindings & body]
  `(if-some* ~bindings (do ~@body)))
```

**Pull request merged**